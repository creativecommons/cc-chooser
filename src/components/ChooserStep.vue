<template>
    <div class="step-content">
        <div
            v-if="status==='previous'||showDisabledDue"
            class="step-description vocab-body body-normal"
        >
            <p class="vocab-body body-normal">
                {{ $t(cardText) }}
            </p>
        </div>
        <div
            v-else-if="status==='current'"
            class="step-actions"
        >
            <div
                class="field"
                :class="yesSelected"
            >
                <b-radio
                    v-model="radio"
                    native-value="yes"
                >
                    <span class="vocab-body body-normal">{{ $t('stepper.yes') }}{{ $t(yesText) }}</span>
                </b-radio>
            </div>
            <div
                class="field"
                :class="noSelected"
            >
                <b-radio
                    v-model="radio"
                    native-value="no"
                >
                    <span class="vocab-body body-normal">{{ $t('stepper.no') }}{{ $t(noText) }}</span>
                </b-radio>
            </div>
        </div>
    </div>
</template>

<script>

export default {
    name: 'ChooserStep',
    props: {
        stepName: String,
        selected: Boolean,
        stepId: Number,
        status: {
            type: String,
            validator(value) {
                return ['current', 'previous', 'inactive'].includes(value)
            }
        },
        reversed: Boolean,
        enabled: Boolean,
        disabledDue: String
    },
    computed: {
        showDisabledDue() {
            return !this.$props.enabled && this.$props.disabledDue !== undefined
        },
        qualifier() {
            return this.reversed ? !this.selected : this.selected
        },
        cardText() {
            console.log('Will return card text: ', this.stepName, this.$props.selected)
            if (this.stepName === 'FS') {
                return this.$props.selected ? 'stepper.FS.selected' : 'stepper.FS.not-selected'
            }
            if (this.$props.enabled === false) {
                return this.$props.disabledDue === 'ND'
                    ? 'stepper.disabled-text-ND'
                    : 'stepper.disabled-text'
            }
            const prefix = `stepper.${this.$props.stepName}.${this.qualifier ? '' : 'not-'}`
            return `${prefix}selected`
        },
        radio: {
            get() {
                if (this.$props.selected === undefined) {
                    return undefined
                } else {
                    return this.qualifier ? 'yes' : 'no'
                }
            },
            set(newVal) {
                let isSelected = newVal === 'yes'
                isSelected = this.reversed ? !isSelected : isSelected
                this.$emit('change', this.$props.stepName, this.$props.stepId, isSelected)
            }
        },
        tPrefix() {
            return 'stepper.' + this.stepName
        },
        question() {
            return this.tPrefix + '.question'
        },
        yesText() {
            return this.tPrefix + '.selected'
        },
        noText() {
            return this.tPrefix + '.not-selected'
        },
        yesSelected() {
            return this.selected ? 'selected' : 'not-selected'
        },
        noSelected() {
            return !this.selected ? 'selected' : 'not-selected'
        }
    }
}
</script>
