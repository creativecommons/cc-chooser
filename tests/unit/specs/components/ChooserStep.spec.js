import { mount, createLocalVue } from '@vue/test-utils'
import ChooserStep from '@/src/components/ChooserStep'
import VueVocabulary from '@creativecommons/vue-vocabulary/vue-vocabulary.common'

describe('ChooserStep.vue', () => {
    let wrapper
    let localVue

    // Always creates a shallow instance of component
    beforeEach(() => {
        localVue = createLocalVue()
        localVue.use(VueVocabulary)
        wrapper = mount(ChooserStep, {
            localVue,
            mocks: {
                $t: key => key
            }
        })
    })
    it('Check that ChooserStep renders correctly', async () => {
        await wrapper.setProps({
            id: 1,
            name: 'BY',
            selected: false
        })
        const yesField = wrapper.findAll('.v-radio').at(0)
        expect(yesField.text()).toBe('stepper.yes'+'stepper.BY.selected')
        const noField = wrapper.findAll('.v-radio').at(1)
        expect(noField.text()).toBe('stepper.no'+'stepper.BY.not-selected')
    })

    it('Check that selecting "Yes" emits correct value', async () => {
        await wrapper.setProps({
            id: 1,
            name: 'BY',
            selected: undefined
        })
        const yesField = wrapper.findAll('.v-radio').at(0)
        expect(yesField.text()).toBe('stepper.yes'+'stepper.BY.selected')
        const noField = wrapper.findAll('.v-radio').at(1)
        expect(noField.text()).toBe('stepper.no'+'stepper.BY.not-selected')
        const radioYes = wrapper.find('input[value="yes"]')
        await radioYes.setChecked()
        expect(wrapper.emitted().change[0]).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    name: 'BY',
                    id: 1,
                    selected: true
                })
            ])
        )
    })

    it('Check reversed step emits reversed value', async () => {
        await wrapper.setProps({
            id: 3,
            name: 'ND',
            selected: undefined
            }
        )
        const radioYes = wrapper.find('input[value="yes"]')
        await radioYes.setChecked()
        expect(wrapper.emitted().change[0]).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    selected: false
                })
            ])
        )
    })
})
