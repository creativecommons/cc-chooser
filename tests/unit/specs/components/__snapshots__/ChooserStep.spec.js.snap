// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooserStep.vue Check if the Step.vue component has the expected UI if current 1`] = `
<div class="step-content">
  <div class="step-actions">
    <div class="field not-selected">
      <b-radio-stub value="no" nativevalue="yes"><span class="vocab-body body-normal">stepper.yesstepper.SA.selected</span></b-radio-stub>
    </div>
    <div class="field selected">
      <b-radio-stub value="no" nativevalue="no"><span class="vocab-body body-normal">stepper.nostepper.SA.not-selected</span></b-radio-stub>
    </div>
  </div>
</div>
`;

exports[`ChooserStep.vue Check if the Step.vue component has the expected UI if disabled due is set 1`] = `
<div class="step-content">
  <div class="step-description vocab-body body-normal">
    <p class="vocab-body body-normal">
      stepper.disabled-text-ND
    </p>
  </div>
</div>
`;

exports[`ChooserStep.vue Check if the Step.vue component has the expected UI if inactive 1`] = `
<div class="step-content">
  <!---->
</div>
`;

exports[`ChooserStep.vue Check if the Step.vue component has the expected UI if previous 1`] = `
<div class="step-content">
  <div class="step-description vocab-body body-normal">
    <p class="vocab-body body-normal">
      stepper.SA.not-selected
    </p>
  </div>
</div>
`;

exports[`ChooserStep.vue Check if the Step.vue component has the expected UI if reversed 1`] = `
<div class="step-content">
  <div class="step-actions">
    <div class="field not-selected">
      <b-radio-stub value="yes" nativevalue="yes"><span class="vocab-body body-normal">stepper.yesstepper.SA.selected</span></b-radio-stub>
    </div>
    <div class="field selected">
      <b-radio-stub value="yes" nativevalue="no"><span class="vocab-body body-normal">stepper.nostepper.SA.not-selected</span></b-radio-stub>
    </div>
  </div>
</div>
`;

exports[`ChooserStep.vue Mark up is correctly rendered 1`] = `
<div
  class="step-content"
>
  <div
    class="step-actions"
  >
    <div
      class="field not-selected"
    >
      <b-radio-stub
        nativevalue="yes"
      >
        <span
          class="vocab-body body-normal"
        >
          stepper.yesstepper.undefined.selected
        </span>
      </b-radio-stub>
    </div>
     
    <div
      class="field selected"
    >
      <b-radio-stub
        nativevalue="no"
      >
        <span
          class="vocab-body body-normal"
        >
          stepper.nostepper.undefined.not-selected
        </span>
      </b-radio-stub>
    </div>
  </div>
</div>
`;
